// @wac/js-harmony/lib/looper.js
function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

/**
 * 轮询HarmonyServer
 */
const remoteRequest = require('./remote-request');
const Util = require('./util');

module.exports = class Looper {
  constructor(harmonyClient) {
    this.harmonyServerIPTimer = null;
    this.harmonyClient = harmonyClient;
    this.hasReady = false;
  }

  /**
   * 启动轮询
   */
  start() {
    this.getHarmonyServerIPList();
  }

  /**
   * 轮询刷新harmony的IPList，间隔5分钟
   */
  getHarmonyServerIPList() {
    var _this = this;

    return _asyncToGenerator(function* () {
      _this.harmonyServerIPTimer && clearTimeout(_this.harmonyServerIPTimer);
      _this.harmonyServerIPTimer = null;

      const res = yield remoteRequest.getHarmonyServerIPList();

      if (res && res.code === 0 && res.data && res.data.length) {
        const data = Util.descendSortByWeight(res.data);

        // 缓存
        Util.setHarmonyIP(data, _this.harmonyClient.cache);
        // 发送数据准备完成通知
        if (!_this.hasReady) {
          _this.hasReady = true;
          _this.harmonyClient.emit('ready');
        }
      } else {
        _this.harmonyServerIPTimer && clearTimeout(_this.harmonyServerIPTimer);
        _this.harmonyServerIPTimer = null;
        _this.getHarmonyServerIPList();
      }

      // 5分钟轮询
      _this.harmonyServerIPTimer = setTimeout(function () {
        _this.getHarmonyServerIPList();
      }, 5 * 60 * 1000);
    })();
  }
};
